redhat-appstudio/service-provider-integration-operator
# add GitLab as supported service provider

### What does this PR do?
Adds GitLab as a new supported service provider.

Some caveats:
-  `read_user` scope is needed by default to fetch user metadata with token
- ServiceProviderState, used to figure out to which repositories the token has access to, will be implemented in its own PR https://issues.redhat.com/browse/SVPI-255
- improved ServiceProvider probing will be implemented in it's own PR after support for multiple service provider user oauth configurations (https://issues.redhat.com/browse/SVPI-249) https://issues.redhat.com/browse/SVPI-256

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

Implements https://issues.redhat.com/browse/SVPI-234

### How to test this PR?

Using OAuth flow: 
1. Deploy SPI with images: `quay.io/pbaran/spis:GITLAB` for oauth service and `quay.io/pbaran/spio:GITLAB` for operator
2. Create a private repository on some instance of GitLab (you can use https://gitlab.com)
3. Create an authorized app in your user settings, like in the picture below, where the redirect URI is your OAuth service URL + /gitlab/callback
![image](https://user-images.githubusercontent.com/73115616/197499144-6df538bf-9957-4f3a-b177-61f7788c1769.png)
4. Copy the client ID (a.k.a. application ID), Secret, and create a new GitLab entry in spi config like the example below:
```yaml
serviceProviders:
- type: GitHub
  clientId: "123"
  clientSecret: "42"
- type: GitLab
  clientId: "your client id..."
  clientSecret: "your client secret..."
  baseUrl: "https://gitlab.com"
```
5. Make sure that OAuth service and operator have loaded the new config.
6. Create a binding such as `samples/binding-write-repo-gitlab.yaml` where you replace the repoUrl with your own.
7. Copy OAuth URL from binding, login, and go through the OAuth flow.
8. SPIAccessToken should then look something like this: 
![image](https://user-images.githubusercontent.com/73115616/197501470-079e21ca-d8c1-4ae8-967d-318efe538718.png)

Using manual upload: The PR should work with GitLab's personal access tokens (PAT) by uploading the token on OAuth service: 
1. Same as 1. and 2. step in OAuth flow.
2. Create PAT in GitLab.
![image](https://user-images.githubusercontent.com/73115616/199254142-00752265-cbdf-4fcc-a6b7-8cfbc704c000.png)
3. Create binding same as in OAuth flow step 6.
4. Copy the token and upload it through token upload URL in binding.
5. SpiAccessToken should look the same as in OAuth flow step 8.
